include/sparrow_v01/layout/list_layout/list_array.hpp

```diff
@@ -0,0 +1,142 @@
+// limitations under the License.
+
+#pragma once
+
+#include "sparrow_v01/layout/array_base.hpp"
+#include "sparrow/layout/layout_iterator.hpp"
+#include "sparrow_v01/layout/nested_value_types.hpp"
+#include "sparrow/utils/iterator.hpp"
+#include "sparrow/utils/nullable.hpp"
+#include "sparrow/array/data_traits.hpp"
+#include "sparrow_v01/array_factory.hpp"
+#include "sparrow_v01/utils/memory.hpp"
+#include "sparrow_v01/utils/functor_index_iterator.hpp"
+#include "sparrow_v01/layout/layout_utils.hpp"
+
+namespace sparrow
+{
+    template <bool BIG>
+    class list_array_impl;
+
+    using list_array = list_array_impl<false>;
+    using big_list_array = list_array_impl<true>;
+
+    template <bool BIG>
+    struct array_inner_types<list_array_impl<BIG>> : array_inner_types_base
+    {
+        using array_type = list_array_impl<BIG>;
+        using inner_value_type = list_value2;
+        using inner_reference  = list_value2;
+        using inner_const_reference = list_value2;
+        using value_iterator = functor_index_iterator<detail::LayoutValueFunctor<array_type>>;
+        using const_value_iterator = functor_index_iterator<detail::LayoutValueFunctor<const array_type>>;
+        using iterator_tag = std::random_access_iterator_tag;
+    };
+
+    template <bool BIG>
+    class list_array_impl final : public array_base,
+                                  public array_crtp_base<list_array_impl<BIG>>
+    {
+        public:
+        using self_type = list_array_impl<BIG>;
+        using base_type = array_crtp_base<self_type>;
+        using inner_types = array_inner_types<self_type>;
+        using value_iterator = typename inner_types::value_iterator;
+        using const_value_iterator = typename inner_types::const_value_iterator;
+        using size_type = typename base_type::size_type;
+        
+        using bitmap_type = typename base_type::bitmap_type;
+        using bitmap_reference = typename base_type::bitmap_reference;
+        using bitmap_const_reference = typename base_type::bitmap_const_reference;
+
+        using inner_value_type =  list_value2;
+        using inner_reference =  list_value2;
+        using inner_const_reference =  list_value2;
+
+
+        using value_type = nullable<inner_value_type>;
+        using reference = nullable<inner_reference, bitmap_reference>;
+        using const_reference = nullable<inner_const_reference, bitmap_const_reference>;
+        using iterator_tag = std::contiguous_iterator_tag;
+
+
+        using flat_array_offset_type = std::conditional_t<BIG, std::uint64_t, std::uint32_t>;
+        using list_size_type = flat_array_offset_type;
+        using flat_array_offset_pointer = flat_array_offset_type*;
+
+
+        explicit list_array_impl(arrow_proxy proxy)
+        :   array_base(proxy.data_type()),
+            base_type(std::move(proxy)),
+            p_list_offsets(reinterpret_cast<flat_array_offset_type*>(this->storage().buffers()[1].data() + this->storage().offset())),
```

Comment:
Can you replace the magic number buffers index [1] with a variable ?

Reply:
sure